caddyhttp/basicauth/basicauth.go

```diff
@@ -63,7 +63,7 @@ func (a BasicAuth) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error
 
 			// let upstream middleware (e.g. fastcgi and cgi) know about authenticated
 			// user; this replaces the request with a wrapped instance
-			r = r.WithContext(context.WithValue(r.Context(),
+			*r = *r.WithContext(context.WithValue(r.Context(),
```

Comment:
@jamesboswell Thanks for the explanation - but it doesn't seem right. The method returns a _copy_ of the original request with the new value in the context, hence the re-assignment of `r`.

@tobya I do remember that issue, but that was different because we had to modify the Request in a function outside of ServeHTTP; in this function, there's no use of a replacer, so why is the dereference necessary?

I'm worried that doing the dereference could easily introduce race conditions in the event a handler ever starts a goroutine (I think some already do) and needs to use the context...